lib/matplotlib/axes/_axes.py

```diff
@@ -5676,24 +5675,63 @@ def _pcolorargs(funcname, *args, allmatch=False):
             raise TypeError(
                 'Incompatible X, Y inputs to %s; see help(%s)' % (
                 funcname, funcname))
-        if allmatch:
-            if (Nx, Ny) != (ncols, nrows):
+
+        if shading == 'auto':
+            if ncols == Nx and nrows == Ny:
+                shading = 'nearest'
+            else:
+                shading = 'flat'
+
+        if shading == 'flat':
+            if not (ncols in (Nx, Nx - 1) and nrows in (Ny, Ny - 1)):
                 raise TypeError('Dimensions of C %s are incompatible with'
                                 ' X (%d) and/or Y (%d); see help(%s)' % (
                                     C.shape, Nx, Ny, funcname))
-        else:
-            if not (ncols in (Nx, Nx - 1) and nrows in (Ny, Ny - 1)):
+            if (ncols == Nx or nrows == Ny):
+                cbook.warn_deprecated("3.3",
+                    message="shading='flat' when X and Y have the same "
+                            "dimensions as C is deprecated since %(since)s.  "
+                            "Either specify the corners of the quadrilaterals "
+                            "with X and Y, or pass shading='auto', 'nearest' "
+                            "or 'gouraud', or set rcParams['pcolor.shading']")
+            C = C[:Ny - 1, :Nx - 1]
+        else:    # ['nearest', 'gouraud']:
+            if (Nx, Ny) != (ncols, nrows):
                 raise TypeError('Dimensions of C %s are incompatible with'
                                 ' X (%d) and/or Y (%d); see help(%s)' % (
                                     C.shape, Nx, Ny, funcname))
-            C = C[:Ny - 1, :Nx - 1]
+            if shading in ['nearest', 'auto']:
+                # grid is specified at the center, so define corners
+                # at the midpoints between the grid centers and then use the
+                # flat algorithm.
+                def _interp_grid(X):
+                    # helper for below
+                    if np.shape(X)[1] > 1:
+                        dX = np.diff(X, axis=1)/2.
+                        X = np.hstack((X[:, [0]] - dX[:, [0]],
+                                       X[:, :-1] + dX,
+                                       X[:, [-1]] + dX[:, [-1]]))
```

Comment:
I guess the whole thing would read slightly nicer if you interpolated rows instead of columns (and then below did the double transpose in the `ncols == Nx` case and not in the `nrows = Ny` case):
```
if len(X) > 1:
    dX = np.diff(X, axis=0) / 2
    X = np.row_stack([X[0] - dX[0], X[:-1] + dX, X[-1] + dX[-1]])
else:
    X = np.row_stack([X, X])
```
(untested)

Reply:
That only reads nicer if you can remember the rule as to whether `X[a]` refers to that a-th row or column.  I prefer the more explicit form.

Reply:
I think it's not too hard if you think of X as "list of lists" (then X[a] is just the a-th list, which is clearly the a-th row).  Anyways, not insisting on it, this is fine too.

Reply:
> which is clearly the a-th row

Why is that clear?  

I think when possible, its helpful to be explicit that we are using a 2-D array versus a 1-D.  If I read `X[0]-dX[0]` I immediately think `X` is 1-D.

Reply:
Because
```
[[1, 2, 3],
 [4, 5, 6]]
```
is just
```
[blah,  # row
 blah]  # row
```
?  i.e. literally how numpy writes out the thing?
Anyways, that's a tangent, not really relevant here.

Reply:
Yeah, its probably just that I have to switch between python and Fortran too much to keep it straight.